File: baseunits/modules/WieManga.pas
unit WieManga;

{$mode objfpc}{$H+}

interface

uses
  Classes, SysUtils, WebsiteModules, uData, uBaseUnit, uDownloadsManager,
  XQueryEngineHTML, httpsendthread, MangaFoxWatermark, FMDVars;

implementation

function GetInfo(const MangaInfo: TMangaInformation; const AURL: String;
  const Module: TModuleContainer): Integer;
begin
  Result := NET_PROBLEM;
  if MangaInfo = nil then Exit(UNKNOWN_ERROR);
  with MangaInfo.mangaInfo do
  begin
    url := MaybeFillHost(Module.RootURL, AURL);
    if MangaInfo.FHTTP.GET(url) then
    begin
      Result := NO_ERROR;
      with TXQueryEngineHTML.Create(MangaInfo.FHTTP.Document) do
        try
          title := XPathString('//div[@class="bookmessagebox"]/h1/substring-before(., " Manga")');
          coverLink := XPathString('//div[@class="bookfrontpage"]/a/img/@src');
          summary := XPathString('//h4[text()="Beschreibung"]/following-sibling::text()');
          genres := XPathStringAll('//div[@class="bookmessgae"]//dd[contains(span/text(), "Genre")]/a');
          authors := XPathStringAll('//div[@class="bookmessgae"]//dd[contains(span/text(), "Autor")]/a');
          artists := XPathStringAll('//div[@class="bookmessgae"]//dd[contains(span/text(), "Zeichner")]/a');
          status := MangaInfoStatusIfPos(XPathString('//div[@class="bookmessgae"]//dd[contains(span/text(), "Status")]/a'), 'ongoing', 'finished');
          XPathHREFAll('//div[@class="chapterlist"]/table//td[@class="col1"]/a', chapterLinks, chapterName);
          InvertStrings([chapterLinks, chapterName]);
        finally
          Free;
        end;
    end;
  end;
end;

function GetPageNumber(const DownloadThread: TDownloadThread; const AURL: String;
  const Module: TModuleContainer): Boolean;
var
  url: String;
begin
  Result := False;
  if DownloadThread = nil then Exit;
  url := MaybeFillHost(Module.RootURL, AURL);
  url := TrimRightChar(url, ['/']) + '-1.html';
  with DownloadThread.FHTTP, DownloadThread.Task.Container do begin
    PageLinks.Clear;
    PageNumber := 0;
    if GET(url) then begin
      Result := True;
      PageNumber := XPathCount('(//select[@id="page"])[1]/option', Document);
    end;
  end;
end;

function GetImageURL(const DownloadThread: TDownloadThread; const AURL: String;
  const Module: TModuleContainer): Boolean;
var
  baseurl, url, s: String;
begin
  Result := False;
  if DownloadThread = nil then Exit;
  baseurl := MaybeFillHost(Module.RootURL, AURL);
  url := TrimRightChar(baseurl, ['/']) + '-' + IncStr(DownloadThread.WorkId) + '.html';
  with DownloadThread.Task.Container, DownloadThread.FHTTP do begin
    Headers.Values['Referer'] := baseurl;
    if GET(url) then begin
      Result := True;
      s := XPathString('//img[@id="comicpic"]/@src', Document);
      PageLinks[DownloadThread.WorkId] := s;
    end;
  end;
end;

function GetNameAndLink(const MangaInfo: TMangaInformation;
  const ANames, ALinks: TStringList; const AURL: String;
  const Module: TModuleContainer): Integer;
var
  s: String;
  v: IXQValue;
  i, x: Integer;
begin
  Result := NET_PROBLEM;
  if Module.CurrentDirectoryIndex = 0 then
    s := '0-9'
  else
    s := ALPHA_LIST_UP[Module.CurrentDirectoryIndex + 1];
  if MangaInfo.FHTTP.GET(Module.RootURL + '/category/' + s + '_'+ IncStr(AURL) + '.html') then
  begin
    Result := NO_ERROR;
    with TXQueryEngineHTML.Create(MangaInfo.FHTTP.Document) do try
      i := 1;
      for v in XPath('//*[@class="booklist"]//span[@class="pagetor"]//text()') do begin
        x := StrToIntDef(v.toString, 1);
        if x > i then i := x;
      end;
      updateList.CurrentDirectoryPageNumber := i;
      XPathHREFtitleAll('//*[@class="booklist"]/table//dl/dd/a[1]', ALinks, ANames);
    finally
      Free;
    end;
  end;
end;

procedure RegisterModule;
begin
  with AddModule do
  begin
    Website := 'WieManga';
    RootURL := 'https://www.wiemanga.com';
    OnGetNameAndLink := @GetNameAndLink;
    OnGetInfo := @GetInfo;
    OnGetPageNumber := @GetPageNumber;
    OnGetImageURL := @GetImageURL;
    TotalDirectory := Length(ALPHA_LIST_UP);
  end;
end;

initialization
  RegisterModule;

end.

File: lua/modules/WieManga.lua
local ALPHA_LIST_UP = '#ABCDEFGHIJKLMNOPQRSTUVWXYZ'

function GetInfo()
  mangainfo.url=MaybeFillHost(module.rooturl,url)
  if http.get(mangainfo.url) then
    x=TXQuery.Create(http.Document)
    mangainfo.coverLink=MaybeFillHost(module.rooturl, x.XPathString('//div[@class="bookfrontpage"]/a/img/@src'))
    if module.website == 'WieManga' then
      mangainfo.title=x.XPathString('//div[@class="bookmessagebox"]/h1/substring-before(., " Manga")')
      mangainfo.summary = x.xpathstring('//h4[text()="Beschreibung"]/following-sibling::text()')
      mangainfo.artists = x.XPathStringAll('//div[@class="bookmessgae"]//dd[contains(span/text(), "Zeichner")]/a')
      mangainfo.authors = x.XPathStringAll('//div[@class="bookmessgae"]//dd[contains(span/text(), "Autor")]/a')
      mangainfo.genres = x.XPathStringAll('//div[@class="bookmessgae"]//dd[contains(span/text(), "Genre")]/a')
      mangainfo.status = MangaInfoStatusIfPos(x.xpathstring('//div[@class="bookmessgae"]//dd[contains(span/text(), "Status")]/a'), 'ongoing', 'finished')
    elseif module.website == 'MangaRussia' then
      mangainfo.title=x.XPathString('//div[@class="bookmessagebox"]/h1/substring-after(., "Манга ")')
      mangainfo.summary = x.xpathstring('//h4[text()="Описание"]/following-sibling::text()')
      mangainfo.authors = x.XPathStringAll('//div[@class="bookmessgae"]//dd[contains(span/text(), "Автор")]/a')
      mangainfo.genres = x.XPathStringAll('//div[@class="bookmessgae"]//dd[contains(span/text(), "Жанры")]/a')
      mangainfo.status = MangaInfoStatusIfPos(x.xpathstring('//div[@class="bookmessgae"]//dd[contains(span/text(), "Перевод")]/a'), 'ongoing', 'finished')
    end
    x.XPathHREFAll('//div[@class="chapterlist"]/table//td[@class="col1"]/a', mangainfo.chapterlinks, mangainfo.chapternames)
    InvertStrings(mangainfo.chapterLinks, mangainfo.chapterNames)
    return no_error
  else
    return net_error
  end
end

function GetPageNumber()
  task.pagelinks.clear()
  task.pagenumber=0
  local s=MaybeFillHost(module.rooturl,url):gsub('/$', '') .. '-1.html'
  if http.get(s) then  
    x=TXQuery.Create(http.Document)
    task.pagenumber=x.xpath('(//select[@id="page"])[1]/option').count
    return true
  else
    return false
  end
end

function GetImageURL()
  local baseurl = MaybeFillHost(module.rooturl,url)
  local s = baseurl:gsub('/$','') .. '-' .. tostring(workid+1) .. '.html'
  http.headers.values['Referer'] = baseurl
  if http.get(s) then
    x=TXQuery.Create(http.Document)
    task.pagelinks[workid]=x.xpathstring('//img[@id="comicpic"]/@src')
    return true
  else
    return false
  end
end

function GetNameAndLink()
  local s = '0-9'
  if module.CurrentDirectoryIndex ~= 0 then
    s = ALPHA_LIST_UP:sub(module.CurrentDirectoryIndex+1,module.CurrentDirectoryIndex+1)
  end
  if http.get(module.rooturl..'/category/' .. s .. '_'.. IncStr(url) .. '.html') then
    x=TXQuery.Create(http.Document)
    v=x.xpath('//*[@class="booklist"]//span[@class="pagetor"]//text()')
    local i = 1
    for j=1,v.count do
      v1 = v.get(j)
      local tmp = tonumber(v1.toString)
      if (tmp ~= nil) and (tmp > i) then i = tmp end
    end
    updatelist.CurrentDirectoryPageNumber = i
    x.XPathHREFtitleAll('//*[@class="booklist"]/table//dl/dd/a[1]', links, names)
    return no_error
  else
    return net_problem
  end
end

function AddWebsiteModule(name, url, category)
  local m = NewModule()
  m.category=category
  m.website=name
  m.rooturl=url
  m.totaldirectory = ALPHA_LIST_UP:len()
  m.lastupdated='February 19, 2018'
  m.ongetinfo='GetInfo'
  m.ongetpagenumber='GetPageNumber'
  m.ongetnameandlink='GetNameAndLink'
  m.OnGetImageURL='GetImageURL'
  return m
end

function Init()
  AddWebsiteModule('WieManga', 'https://www.wiemanga.com', 'German')
  AddWebsiteModule('MangaRussia', 'http://www.mangarussia.com', 'Russian')
end 
